runners/direct-java/src/main/java/org/apache/beam/runners/direct/BoundedReadEvaluatorFactory.java

```diff
@@ -126,18 +130,39 @@ public TransformResult finishBundle()  {
     }
 
     @Override
-    public Collection<CommittedBundle<?>> getInitialInputs(AppliedPTransform<?, ?, ?> transform) {
-      return createInitialSplits((AppliedPTransform) transform);
+    public Collection<CommittedBundle<?>> getInitialInputs(
+        AppliedPTransform<?, ?, ?> transform, int targetParallelism) throws Exception {
+      return createInitialSplits((AppliedPTransform) transform, targetParallelism);
     }
 
-    private <OutputT> Collection<CommittedBundle<?>> createInitialSplits(
-        AppliedPTransform<?, ?, Read.Bounded<OutputT>> transform) {
+    private <OutputT>
+        ImmutableList<CommittedBundle<BoundedSourceShard<OutputT>>> createInitialSplits(
+            AppliedPTransform<?, ?, Bounded<OutputT>> transform, int targetParallelism)
+            throws Exception {
       BoundedSource<OutputT> source = transform.getTransform().getSource();
-      return Collections.<CommittedBundle<?>>singleton(
-          evaluationContext
-              .<BoundedSourceShard<OutputT>>createRootBundle()
-              .add(WindowedValue.valueInGlobalWindow(BoundedSourceShard.of(source)))
-              .commit(BoundedWindow.TIMESTAMP_MAX_VALUE));
+      PipelineOptions options = evaluationContext.getPipelineOptions();
+      long estimatedBytes = source.getEstimatedSizeBytes(options);
+      long bytesPerBundle = estimatedBytes / targetParallelism;
+      List<? extends BoundedSource<OutputT>> bundles =
+          source.splitIntoBundles(bytesPerBundle, options);
+      ImmutableList.Builder<CommittedBundle<BoundedSourceShard<OutputT>>> shards =
+          ImmutableList.builder();
+      if (bundles.isEmpty()) {
```

Comment:
This logic shouldn't really exist at all.

If a `Source` produces no splits, the watermark should naturally advance to positive infinity, and the executor should never notice.